### PR TITLE
Fixed reconnection to MySQL

### PR DESCRIPTION
**Pull request by matcracker**
In the case of disconnection from MySQL, libasynql was unable to restore the connection due to the exception thrown.
The following PR has the following fix: in case of a failed ping to the database, the thread tries to restore the connection until successful and continue the execution of the queries in the queue.

I also wanted to propose to set a maximum number of attempts after which to throw an exception or to end the thread execution, what do you think?

### CODE REVIEW

**Comment by SOF3** [comment]
It would be nice to sleep for 2^n seconds for the nth time this loop is executed.

**Comment by SOF3** [comment]
Furthermore, logging connection failure would be preferrable.

**Comment by matcracker** [comment]
So, you suggest me to increase the time of thread sleep every time it fails? Sounds good to me. In case of success the time is reset.

The connection failure is logged by default from mysqli.

**Comment by SOF3** [comment]
You mean mysqli triggers a PHP warning implicitly? We probably want to suppress that with `@`. Those messages won't get logged to server.log.

**Comment by matcracker** [comment]
Right. I was thinking to use MainLogger but I don't like the idea, any other suggestions?

**Comment by matcracker** [comment]
Forget it, I can't use the MainLogger.

**Comment by matcracker** [comment]
Here is a result:
image

**Comment by SOF3** [comment]
Why do you use the server logger? There is the plugin logger...

File: libasynql/src/poggit/libasynql/mysqli/MysqlCredentials.php [hunk 1]
```diff
@@ -104,14 +104,9 @@ public function newMysqli() : mysqli{
...
 	 * @param mysqli $mysqli
-	 *
-	 * @throws SqlError
 	 */
 	public function reconnectMysqli(mysqli &$mysqli) : void{
-		$mysqli->connect($this->host, $this->username, $this->password, $this->schema, $this->port, $this->socket);
-		if($mysqli->connect_error){
-			throw new SqlError(SqlError::STAGE_CONNECT, $mysqli->connect_error);
-		}
+		@$mysqli->connect($this->host, $this->username, $this->password, $this->schema, $this->port, $this->socket);
```
Comment (SOF3) [inline libasynql/src/poggit/libasynql/mysqli/MysqlCredentials.php:111]: I consider using `@` without immediately checking as an antipattern. I know you checked it on line 104 of MysqliThread, but it is too far away. It is too easy to accidentally forget to handle this error in the future if we change calls to this function.

File: libasynql/src/poggit/libasynql/mysqli/MysqliThread.php [hunk 2]
```diff
@@ -82,13 +94,20 @@ protected function executeQuery($mysqli, int $mode, string $query, array $params
...
-		while(!$mysqli->ping()){
-			$cred->reconnectMysqli($mysqli);
-			if($this->connError === null){
-				break;
-			}
+		/** @var AttachableThreadedLogger $logger */
+		$logger = unserialize($this->logger);
+		if(!$mysqli->ping()){
+			do{
+				$seconds = min(2 ** $this->attempts++, PHP_INT_MAX);
```
Comment (SOF3) [inline libasynql/src/poggit/libasynql/mysqli/MysqliThread.php]: This code is confusing. The use of `$x++` here is not idiomatic enough. Please split `++` into a separate statement.

Comment (SOF3) [reply libasynql/src/poggit/libasynql/mysqli/MysqliThread.php]: Btw, capping by `PHP_INT_MAX` can never happen, because in the 31st sleep we need to sleep for 68 years :upside_down_face: But it's OK to leave it here.

File: libasynql/src/poggit/libasynql/mysqli/MysqliThread.php [hunk 3]
```diff
@@ -42,28 +43,38 @@
...
 use function unserialize;
+use const PHP_INT_MAX;
 
 class MysqliThread extends SqlSlaveThread{
 	/** @var string */
 	private $credentials;
+	/** @var string */
+	private $logger;
+	/** @var int */
+	private $attempts;
```
Comment (SOF3) [inline libasynql/src/poggit/libasynql/mysqli/MysqliThread.php]: Is it necessary to store this as a class property? It seems you are only using it as a local variable.

File: libasynql/src/poggit/libasynql/mysqli/MysqliThread.php [hunk 4]
```diff
@@ -138,22 +159,26 @@ protected function executeQuery($mysqli, int $mode, string $query, array $params
 				case SqlThread::MODE_GENERIC:
 					$ret = new SqlResult();
 					$stmt->close();
+
```
Comment (SOF3) [inline libasynql/src/poggit/libasynql/mysqli/MysqliThread.php]: Please don't mix unrelated formatting changes in the same pull request.

File: libasynql/src/poggit/libasynql/base/SqlSlaveThread.php [hunk 5]
```diff
@@ -96,7 +97,7 @@ public function run() : void{
 	/**
 	 * @return bool
 	 */
-	public function isBusy(): bool {
+	public function isBusy() : bool{
```
Comment (SOF3) [inline libasynql/src/poggit/libasynql/base/SqlSlaveThread.php]: This change is irrelevant to this pull request.

Consider setting "Only VCS changed text" in your phpstorm settings.

How spaces are placed here is not a big deal since they do not affect readability a lot (just OCD). But it is a problem if the git blame is corrupted by irrelevant changes.

File: libasynql/src/poggit/libasynql/mysqli/MysqliThread.php [hunk 6]
```diff
@@ -42,28 +43,38 @@
...
+	public static function createFactory(MysqlCredentials $credentials, AttachableThreadedLogger $logger) : Closure{
+		return function(SleeperNotifier $notifier, QuerySendQueue $bufferSend, QueryRecvQueue $bufferRecv) use ($credentials, $logger){
+			return new MysqliThread($credentials, $notifier, $logger, $bufferSend, $bufferRecv);
 		};
 	}
 
-	public function __construct(MysqlCredentials $credentials, SleeperNotifier $notifier, QuerySendQueue $bufferSend = null, QueryRecvQueue $bufferRecv = null){
+	public function __construct(MysqlCredentials $credentials, SleeperNotifier $notifier, AttachableThreadedLogger $logger, QuerySendQueue $bufferSend = null, QueryRecvQueue $bufferRecv = null){
 		$this->credentials = serialize($credentials);
+		$this->logger = serialize($logger);
```
Comment (SOF3) [inline libasynql/src/poggit/libasynql/mysqli/MysqliThread.php]: Are you sure we need to `serialize` this? `$logger` is `Threaded`.

**Review by SOF3 — CHANGES_REQUESTED** [review]

**Comment by matcracker** [comment]
> Why do you use the server logger? There is the plugin logger...

Because PluginLogger extends AttachableLogger and not AttachableThreadedLogger.

File: libasynql/src/poggit/libasynql/mysqli/MysqlCredentials.php (hunk 1 above)
Comment (matcracker) [reply libasynql/src/poggit/libasynql/mysqli/MysqlCredentials.php:111]: Here is a problem. I revert this change except for `@`, so it thows the SqlError. In the MysqliThread I changed in the following way:
image

The problem is...
image

Comment (matcracker) [reply libasynql/src/poggit/libasynql/mysqli/MysqlCredentials.php:111]: In this way it works
image

**Comment by matcracker** [comment]
I just thought that in case of more than one disconnected worker, each one will try to reconnect and consequently there will be a spam of messages. I should think another way, do you?

**Comment by SOF3** [comment]
> I just thought that in case of more than one disconnected worker, each one will try to reconnect and consequently there will be a spam of messages. I should think another way, do you?

I don't think that's a problem. The reconnection status of each thread is independent, and the logger detects the current thread name anyway.

In case of shaded virions, the thread name contains the shaded namespace, so it should be clear which plugin it came from.

**Comment by dadodasyra** [comment]
This would be awesome

**Comment by SOF3** [comment]
please rebase the PR, looks good to me

**Comment by matcracker** [comment]
Done!

**Review by SOF3 — APPROVED** [review]
